hoglet.go

```diff
@@ -3,114 +3,170 @@ package hoglet
+// A Circuit with a nil breaker will never open.
+func NewCircuit[IN, OUT any](f BreakableFunc[IN, OUT], breaker Breaker, opts ...Option) *Circuit[IN, OUT] {
+	b := &Circuit[IN, OUT]{
 		f:       f,
-		trigger: trigger,
+		breaker: breaker,
 		options: options{
 			isFailure: defaultFailureCondition,
 		},
 	}
 
+	if breaker != nil {
+		breaker.connect(b)
+	}
+
 	for _, opt := range opts {
 		opt.apply(&b.options)
 	}
 
 	return b
 }
 
-// Do calls the wrapped function and returns its result. If the breaker is open, it returns [ErrBreakerOpen].
+// State reports the current [State] of the circuit.
+// It should only be used for informational purposes. To minimize race conditions, the circuit should be called directly
+// instead of checking its state first.
+func (c *Circuit[IN, OUT]) State() State {
+	oa := c.openedAt.Load()
+
+	if oa == 0 {
+		// closed
+		return StateClosed
+	}
+
+	if c.halfOpenDelay == 0 || time.Since(time.UnixMicro(oa)) < c.halfOpenDelay {
+		// open
+		return StateOpen
+	}
+
+	// half-open
+	return StateHalfOpen
+}
+
+// stateForCall returns the state of the circuit meant for the next call.
+// It wraps [State] to keep the mutable part outside of the external API.
+func (c *Circuit[IN, OUT]) stateForCall() State {
+	state := c.State()
+
+	if state == StateHalfOpen {
+		// We reset openedAt to block further calls to pass through when half-open. A success will cause the breaker to
+		// close. This is slightly racy: multiple goroutines may reach this point concurrently since we do not lock the
+		// breaker.
+		// CompareAndSwap is needed to avoid clobbering another goroutine's openedAt value.
+		c.setOpenedAt(time.Now().UnixMicro())
+	}
+
+	return state
+}
+
+// SetOpenedAt sets the time the circuit was opened at.
+// Passing a value of 0 (re)opens the cirtuit.
+func (c *Circuit[IN, OUT]) setOpenedAt(i int64) {
+	if i == 0 {
+		c.openedAt.Store(i)
+	} else {
+		c.openedAt.CompareAndSwap(0, i)
+	}
+}
+
+// Call calls the wrapped function if the circuit is closed and returns its result. If the circuit is open, it returns
+// [ErrCircuitOpen].
 //
 // The wrapped function is called synchronously, but possilble context errors are recorded as soon as they occur. This
-// ensures the breaker opens quickly, even if the wrapped function blocks.
+// ensures the circuit opens quickly, even if the wrapped function blocks.
 //
 // By default, all errors are considered failures (including [context.Canceled]), but this can be customized via
 // [WithFailureCondition] and [IgnoreContextCancelation].
 //
-// Panics are observed as failures, but are not recovered.
-func (b *Breaker[IN, OUT]) Do(ctx context.Context, in IN) (out OUT, err error) {
-	if b.f == nil {
+// Panics are observed as failures, but are not recovered (i.e.: they are "repanicked" instead).
+func (c *Circuit[IN, OUT]) Call(ctx context.Context, in IN) (out OUT, err error) {
+	if c.f == nil {
```

Comment:
Should we do this nil check in NewCircuit?

Reply:
I had this here to make the zero value for `Circuit` usable without needing an `error` response, but on second thought if you have a nil function, your code will panic, so I think it's perfectly fine if you wrap a nil function, your code will still panic. 

I've removed the check and the test. PTAL.